### PR TITLE
Support multiple memories in RemoveUnusedModuleElements

### PR DESCRIPTION
**Pull request by tlively**
Add support for memory and data segment module elements and treat them uniformly
with other module elements rather than as special cases. There is a cyclic
dependency between memories (or tables) and their active segments because
exported or accessed memories (or tables) keep their active segments alive, but
active segments for imported memories (or tables) keep their memories (or
tables) alive as well.

### CODE REVIEW

**Comment by tlively** [comment]
Current dependencies on/for this PR:
* main
  * **PR #5601** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/5601" target="_blank">Graphite</a> 
    * **PR #5602** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/5602" target="_blank">Graphite</a> 
      * **PR #5603** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/5603" target="_blank">Graphite</a> 
        * **PR #5604** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/5604" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/WebAssembly/binaryen/5604?utm_source=stack-comment).

File: src/passes/RemoveUnusedModuleElements.cpp [hunk 1]
```diff
@@ -643,20 +681,26 @@ struct RemoveUnusedModuleElements : public Pass {
...
+        roots.emplace_back(ModuleElementKind::Memory, curr->value);
       }
     }
-    // Check for special imports, which are roots.
-    bool importsMemory = false;
-    if (!module->memories.empty() && module->memories[0]->imported()) {
-      importsMemory = true;
-    }
+
+    // Active segments that write to imported tables and memories are roots.
```
Comment (kripken) [inline src/passes/RemoveUnusedModuleElements.cpp:689]: Why are these roots? The import may not be used, in which case neither are they.

File: src/passes/RemoveUnusedModuleElements.cpp [hunk 2]
```diff
@@ -702,6 +746,24 @@ struct RemoveUnusedModuleElements : public Pass {
...
+    });
+    module->removeMemories([&](Memory* curr) {
+      return !needed({ModuleElementKind::Memory, curr->name});
+    });
+    module->removeTables([&](Table* curr) {
+      return !needed({ModuleElementKind::Table, curr->name});
+    });
+    module->removeDataSegments([&](DataSegment* curr) {
+      return !needed({ModuleElementKind::DataSegment, curr->name});
+    });
```
Comment (kripken) [inline src/passes/RemoveUnusedModuleElements.cpp:766]: Looks like the first and last 3 here are duplicates? The first three may have been intended to be deleted?

File: src/passes/RemoveUnusedModuleElements.cpp (hunk 1 above)
Comment (tlively) [reply src/passes/RemoveUnusedModuleElements.cpp:689]: Even if the module never uses the imports, the fact that the segments are written to the imported table or memory during initialization is externally visible, so we have to make sure it still happens after the optimization.

Comment (kripken) [reply src/passes/RemoveUnusedModuleElements.cpp:689]: Oh, good point...

**Review by kripken — APPROVED** [review]
lgtm minus that one 3/3 fix.

**Comment by tlively** [comment]
The fuzzer found a bug where removing segments produces invalid modules because the segment indices in various instructions aren't updated. I've wanted to replace the indexes with names to match all other module elements for a long time, and now I finally have a reason to do that 🎉

**Comment by tlively** [comment]
This depends on #5618 to resolve that fuzzer bug.

**Comment by tlively** [comment]
Rebased and fuzzing again.

**Pull request merged**